python/newrelic_lambda/lambda_handler.py

```diff
@@ -0,0 +1,274 @@
+            return None
+        except KeyError:
+            return None
+    return pos
+
+
+def extract_event_source_arn(event):
+    try:
+        # Firehose
+        arn = event.get("streamArn") or event.get("deliveryStreamArn")
+
+        if not arn:
+            # Dynamo, Kinesis, S3, SNS, SQS
+            record = path_get("Records[0]", event)
+
+            if record:
+                arn = (
+                    record.get("eventSourceARN")
+                    or record.get("EventSubscriptionArn")
+                    or path_get("s3.bucket.arn", record)
+                )
+        # ALB
+        if not arn:
+            arn = path_get("requestContext.elb.targetGroupArn", event)
+        # CloudWatch events
+        if not arn:
+            arn = path_get("resources[0]", event)
+
+        if arn:
+            return newrelic.core.attribute.truncate(str(arn))
+        return None
+    except Exception:
+        pass
+
+
+def detect_event_type(event):
+    if isinstance(event, dict):
+        for k, type_info in EVENT_TYPE_INFO.items():
+            if all(path_match(path, event) for path in type_info["required_keys"]):
+                return type_info
+    return None
+
+
+def get_attributes_for_event_type(event_type, event):
+    attr_names_and_values = {}
+    event_type_attributes = event_type["attributes"]
+    for attr_name, path in event_type_attributes.items():
+        attr = path_get(path, event)
+        if attr is not None:
+            attr_names_and_values[attr_name] = attr
+    return attr_names_and_values
+
+
+def LambdaHandlerWrapper(wrapped, application=None, name=None, group=None):
+    def set_agent_attr(transaction, key, value):
+        # noinspection PyProtectedMember
+        transaction._add_agent_attribute(key, value)
+
+    def _nr_lambda_handler_wrapper_(wrapped, instance, args, kwargs):
+        # Check to see if any transaction is present, even an inactive
+        # one which has been marked to be ignored or which has been
+        # stopped already.
+
+        transaction = newrelic.agent.current_transaction(active_only=False)
+
+        if transaction:
+            return wrapped(*args, **kwargs)
+
+        try:
+            event, context = args[:2]
+        except Exception:
+            return wrapped(*args, **kwargs)
+
+        target_application = application
+
+        # If application has an activate() method we assume it is an
+        # actual application. Do this rather than check type so that
+        # can easily mock it for testing.
+
+        # FIXME Should this allow for multiple apps if a string.
+
+        if not hasattr(application, "activate"):
+            target_application = newrelic.agent.application(application)
+
+        try:
+            if "httpMethod" in event:
+                request_method = event["httpMethod"]
+                request_path = event["path"]
+            else:
+                request_method = event["requestContext"]["http"]["method"]
```

Comment:
We should probably add safeguard logic here to ensure all of these nested keys are present in the event.